src/golang/lib/workflow/orchestrator/orchestrator.go

```diff
@@ -29,479 +19,201 @@ const (
-				// There is an user error when executing the operator.
-				log.Errorf("Failed due to user error. Operator name %s, id %s", op.Name, op.Id)
-				return ErrOpExecBlockingUserFailure
-			}
+	if orch.shouldPersistResults {
+		err = dag.PersistResult(ctx, status)
+		if err != nil {
+			log.Errorf("Error when persisting dag resutls: %v", err)
+		}
+	}
+	return status, err
+}
 
-			completedIds = append(completedIds, id)
+func waitForInProgressOperators(
+	ctx context.Context,
+	inProgressOps map[uuid.UUID]operator.Operator,
+	pollInterval time.Duration,
+	timeout time.Duration,
+) {
+	start := time.Now()
+	for len(inProgressOps) > 0 {
+		if time.Since(start) > timeout {
+			return
+		}
 
-			for _, artifactId := range op.Outputs {
-				if downstreampOps, ok := artifactToDownstreamOperatorIds[artifactId]; ok {
-					for _, downstreamOpId := range downstreampOps {
-						if _, ok := operatorDependencies[downstreamOpId][artifactId]; !ok {
-							return ErrInvalidArtifactId
-						}
+		for opID, op := range inProgressOps {
+			execState, err := op.GetExecState(ctx)
 
-						delete(operatorDependencies[downstreamOpId], artifactId)
-						if len(operatorDependencies[downstreamOpId]) == 0 {
-							ready[downstreamOpId] = true
-						}
-					}
-				}
+			// Resolve any jobs that aren't actively running or failed. We don't are if they succeeded or failed,
+			// since this is called after orchestration exits.
+			if err != nil || execState.Status != shared.RunningExecutionStatus {
+				delete(inProgressOps, opID)
 			}
 		}
+		time.Sleep(pollInterval)
 	}
+}
 
-	for _, id := range completedIds {
-		delete(active, id)
+func opFailureError(failureType shared.FailureType, op operator.Operator) error {
+	if failureType == shared.SystemFailure {
+		return ErrOpExecSystemFailure
+	} else if failureType == shared.UserFailure {
+		log.Errorf("Failed due to user error. Operator name %s, id %s", op.Name(), op.ID())
+		return ErrOpExecBlockingUserFailure
 	}
-
-	return nil
+	return errors.Newf("Internal error: Unsupported failure type %s", failureType)
 }
 
-func scheduleOperators(
+func execute(
 	ctx context.Context,
-	operators map[uuid.UUID]operator.DBOperator,
-	artifacts map[uuid.UUID]artifact.DBArtifact,
-	ready map[uuid.UUID]bool,
-	active map[uuid.UUID]bool,
-	operatorIdToJobId map[uuid.UUID]string,
-	storageConfig *shared.StorageConfig,
-	artifactContentPaths map[uuid.UUID]string,
-	artifactMetadataPaths map[uuid.UUID]string,
-	operatorMetadataPaths map[uuid.UUID]string,
+	dag dag.WorkflowDag,
+	pollInterval time.Duration,
+	timeout time.Duration,
+	cleanupTimeout time.Duration,
 	jobManager job.JobManager,
-	vaultObject vault.Vault,
+	shouldPersistResults bool,
 ) error {
-	for id := range ready {
-		op, ok := operators[id]
-		if !ok {
-			return ErrInvalidOpId
-		}
+	// These are the operators of immediate interest. They either need to be scheduled or polled on.
+	inProgressOps := make(map[uuid.UUID]operator.Operator, len(dag.Operators()))
```

Comment:
not sure I follow the proposal - what are we sticking on the orchestrator implementation?

Reply:
I'm thinking about making `inProgressOps`, `completedOps` and `status` as a part of `orchestratorImpl` so that any given such object represents a full execution state. We could move this `execute()` to `orchestratorImpl` 's struct method

Reply:
ok, is this so that we can add  a similar GetExecStatus() method on the orchestrator interface?

Reply:
I don't think we have to do that now if there's no external caller need that. We already persist wf using `wf.Persist(status)` which we could still keep.

My motivation is more to clarify the ownership of exec state. One use case, for example, is in future we could simply log `json.Marshal(orchestratorObj)` and that recovered log will contain all information of an intermediate state.

Reply:
We could also discuss in PR review meeting on this